### ISSUE TITLE
Very long start up time with Network I/O blocking

### ISSUE THREAD

**Issue by retry** [mintty/mintty#960]
Windows 10, cygwin 64 bit user here.  I have the default mintty shortcut set up with the target as follow: D:\cygwin\bin\mintty.exe -i /Cygwin-Terminal.ico -

I have no idea what changed, but now it takes half a minute, sometimes longer, for mintty to completely load.  I can see it in task manager almost right away.  When I go to view the wait chain, it says, "One or more threads of mintty.exe are waiting to finish Network I/O". 

I don't know why mintty would be trying to do network I/O at all.  It does not trigger a pop up in my firewall.  I see no evidence of network activity in Tcpview.  There is nothing in the Windows event log to indicate any problem.  I tried to downgrade to the previous mintty (3.0.6) but it's behaving exactly the same way.

Any ideas?

**Comment by k-takata** [comment]
Is that really a mintty issue?
How about executing Cygwin without mintty? (E.g. execute `D:\cygwin\Cygwin.bat`)

**Comment by mintty** [comment]
Symptoms sound like an anti-virus scanner issue. Any newly installed/updated/reconfigured?
Another idea, although not consistent with your description (as there is no network activity observed): Could your mintty be infected? Make a copy, reinstall and compare binaries.

**Comment by BrianInglis** [comment]
Do you see the mintty window quickly? In that case mintty is running and the shell is starting up. 
Are you waiting for the shell prompt to appear? In that case shell initialization scripts are running and something has changed, either in the scripts, or in your system, that may be slowing down or blocking what the scripts are doing. 
Have you installed or upgraded anything recently? See if you can deinstall or downgrade those. 
What else may have changed in your system or environment; is it a home system? In that case, either a patch update or feature update may have been applied. 
Is it a work system? In that case, updates or policy changes may have been applied, and you should talk to your IT support.

**Comment by retry** [comment]
If I just load bash with a bat file, it loads up instantly.  I also tried running some commands directly from windows like bc (opens a shell instantly). The problem is entirely mintty specific.  I have no AVS running at all.  I just run comodo for a firewall.  It's not blocking mintty (no rules set).  If mintty were making a TCP/IP connection, the firewall would alert with a pop up, so I really don't think it's that.  Running mintty from my shortcut (or directly from the cygwin/bin/ dir) is the same result -- nothing appears to happen, but I see a mintty proc appear in task manager.  That is where I can right click and select "Analayze Wait Chain".  It says: "One or more threads of mintty.exe are waiting to finish Network I/O".  Please believe there is nothing else on my system that is doing this. Nothing.

So to sum up responses from comments:
1) Yes it seems to be mintty specific
2) It is not avs because no avs is running (windows defender is off)
3) This is the mintty binary from cygwin.  I tried to downgrade to the earlier version, then upgrade back to the latest version.  No effect.  Hard to believe it's infected.
4) I do not see the mintty window quickly.  I changed the start up string in the shortcut to use bash -xl for debugging and there was nothing there to indicate the delay was in bash.  Remember bash loads instantly from a bat file.
5) It's a home system not connect to an Active Directory or Windows workgroup.  I tried to set up /etc/passwd and /etc/group and set nsswitch.conf to search files only (no db).  This made zero difference.  Mintty is trying to do something and it seems to have nothing to do with cygwin or bash.  Running it as administrator or not makes no difference either.

It would be helpful to understand what network I/O mintty wants to do at startup.  Does it depend on or expect specific windows services to be running?

**Comment by retry** [comment]
This may or may not be related. Why does mintty try to make an https connection to 199.232.28.133 when I open the options dialogue?  That does yield a firewall alert which I can block/allow.  There is nothing seemingly network related in the options dialogue.  Is this just some telemetry beacon or an update check?

**Comment by mintty** [comment]
It's an update check with raw.githubusercontent.com. I was about to ask whether some hotkey tool might send an Alt+space to mintty when it's started, which would open its config menu...
One other case of network access is when you load a colour scheme from a URL by drag-and-drop, also within the Options dialog.

**Comment by mintty** [comment]
Is your %APPDATA% folder located on a network drive? Mintty checks the location for a config file.

**Comment by mintty** [comment]
Please run mintty from a command line (e.g. other mintty) - same behaviour, I assume - and add a `-Rp` option. Is there immediate or delayed output?

**Comment by mintty** [comment]
If `mintty -Rp` delays, please also check `mintty -Rf` and `mintty -Rm`; output immediate or delayed?

**Comment by retry** [comment]
No network drives on my system at all.  I have two SSDs (C and D).  64bit cygwin is installed on D.  I have a ramdrive on R: and user's Temp folder is mounted there.  It has been this way for a while now and never caused problems in the past.

mintty from command line:
1) running mintty from within mintty delays like before
2) mintty -Rp also delays and outputs 
$ mintty -Rp
61294
but that output doesn't show up until the new terminal launches (about 1 minute later)
3) mintty -Rf is immediate but doesn't launch a shell. It just outputs a list of fonts.
4) mintty -Rm is immediate and outputs (also no new shell)
$ mintty -Rm
Monitor 1 current primary (120 dpi) w,h 1920,1080 (   0,   0...1920,1080)
1 monitors,      smallest width,height 1920,1080

**Comment by mintty** [comment]
Did you configure a background image? Can you retest without config file, please?
There isn't much left that would happen between immediate output of -Rm and delayed output of -Rp.
Maybe a forking issue...
Can you compile mintty yourself and try to add some tracing printf calls in winmain.c between "  if (report_moni)" and "  SetFocus(wnd);"?

**Comment by BrianInglis** [comment]
Assuming you installed mintty using Cygwin setup downloaded from the Cygwin site, mintty does no  network access. 
That host is on the Fastly CDN with no reverse DNS PTR. 
Comodo Firewall must get its filtering rules from some CDN. 
Comodo "Firewall" also appears to act as an AV by any other name: 
Comodo Firewall is your first step in keeping hackers, viruses, and malware out of your PC.
Featuring Default Deny Protection™, our Firewall checks an extensive list of over two million known PC-friendly applications. If it's not there, Comodo Firewall lets you know before opening your PC's door.
Firewall Features
Keeps you updated on all suspicious files
Prevention-based technology stops viruses
Default Deny Protection™ so only safe files execute
Easily learns your PC habits for personalized alerts
Automatic updates for the most current protection
also:
Auto Sandbox Technology™ tests programs in a virtual operating environment ensuring viruses & other malicious software are isolated from the rest of your computer.
So if CF does not recognize the version of mintty you are running who knows what it's doing, what it's intercepting, and what it's sending and receiving, especially if you are using the "free" level, as that tends to be monetized by intrusive monitoring. 
If you have not disabled Windows Firewall, it and Comodo Firewall may be blocking or interfering with each other at some points in the network access stacks: see [Cygwin BLODA](https://cygwin.com/faq/faq.html#faq.using.bloda) which lists Comodo Firewall as something to be avoided or deinstalled before using Cygwin.

**Comment by retry** [comment]
This is pure finger pointing and false assumption.  I have been using Comodo firewall with cygwin for something like 15 years.  There are no CDN rules.  The only rules which exist are the ruleset I built up via pop up alerts.  Besides if Mintty does no network activity as you claim, why would it matter?  Windows firewall is disabled via:
netsh advfirewall set allprofiles state off

You cannot disable the service in windows 10.  Besides if I disable Comodo firewall, the problem persists.  This is a completely unhelpful post, which moves us away from diagnosing the root problem: a problem between mintty and windows 10.

Mintty makes a network call if you open the options dialogue. Yes even from the cygwin install.
https://i.imgur.com/r3KoOfA.png

But I don't want to get caught up in all these red herrings.  The options dialogue is not the issue at hand.  What is the issue is this network I/O wait block
https://i.imgur.com/Uiw9laH.png

Finally I'll show you the thread list from a starting up and blocked mintty proc on the left and an already running smoothly mintty proc on the right:
https://i.imgur.com/KUvywB0.png

After a minute or so, the proc on the left will look like the proc on the right.  I don't know anything about windows internals, but just looking for differences, I see combase.dll on the hanging proc, and WINMM.dll running on the good mintty proc.  So could it be combase.dll is blocking or something is holding WINMM.dll from launching?


> Did you configure a background image? 
No I use the green text on black preset

> Can you retest without config file, please?
Sure, if you mean the .minttyrc file I can remove it.  Moving .minttyrc to .minttyrc.save made no difference I'm afraid.

I can probably compile mintty a little later.  Can I do it with cygwin's gcc ?

**Comment by mintty** [comment]
OK, it sounds like Comodo is the culprit then. Yet it would be interesting to narrow down at which point in mintty code Comodo interferes with it. My prime suspicion is fork. I'd appreciate if we can trace that down. It's quite easy to build mintty. Cygwin packages gcc-core and make are sufficient.

**Comment by retry** [comment]
What! No it's NOT.  Please don't finger point the problem away.  I just told you if I disable comodo, it still hangs.

**Comment by retry** [comment]
Are you assuming combase.dll has something to do with Comodo? It does NOT.  It's COM base:

~ $ locate combase.dll 
/cygdrive/c/Windows/servicing/LCU/Package_for_RollupFix~31bf3856ad364e35~amd64~~18362.592.1.5/amd64_microsoft-windows-com-base_31bf3856ad364e35_10.0.18362.449_none_63ca3a92b81babcb/f/combase.dll
/cygdrive/c/Windows/servicing/LCU/Package_for_RollupFix~31bf3856ad364e35~amd64~~18362.592.1.5/amd64_microsoft-windows-com-base_31bf3856ad364e35_10.0.18362.449_none_63ca3a92b81babcb/r/combase.dll
/cygdrive/c/Windows/servicing/LCU/Package_for_RollupFix~31bf3856ad364e35~amd64~~18362.592.1.5/wow64_microsoft-windows-com-base_31bf3856ad364e35_10.0.18362.449_none_6e1ee4e4ec7c6dc6/f/combase.dll
/cygdrive/c/Windows/servicing/LCU/Package_for_RollupFix~31bf3856ad364e35~amd64~~18362.592.1.5/wow64_microsoft-windows-com-base_31bf3856ad364e35_10.0.18362.449_none_6e1ee4e4ec7c6dc6/r/combase.dll
/cygdrive/c/Windows/System32/combase.dll
/cygdrive/c/Windows/SysWOW64/combase.dll
/cygdrive/c/Windows/WinSxS/amd64_microsoft-windows-com-base_31bf3856ad364e35_10.0.18362.449_none_63ca3a92b81babcb/combase.dll
/cygdrive/c/Windows/WinSxS/amd64_microsoft-windows-com-base_31bf3856ad364e35_10.0.18362.449_none_63ca3a92b81babcb/f/combase.dll
/cygdrive/c/Windows/WinSxS/amd64_microsoft-windows-com-base_31bf3856ad364e35_10.0.18362.449_none_63ca3a92b81babcb/r/combase.dll
/cygdrive/c/Windows/WinSxS/Backup/amd64_microsoft-windows-com-base_31bf3856ad364e35_10.0.18362.449_none_63ca3a92b81babcb_combase.dll_a2567a6a
/cygdrive/c/Windows/WinSxS/Backup/wow64_microsoft-windows-com-base_31bf3856ad364e35_10.0.18362.449_none_6e1ee4e4ec7c6dc6_combase.dll_a2567a6a
/cygdrive/c/Windows/WinSxS/wow64_microsoft-windows-com-base_31bf3856ad364e35_10.0.18362.449_none_6e1ee4e4ec7c6dc6/combase.dll
/cygdrive/c/Windows/WinSxS/wow64_microsoft-windows-com-base_31bf3856ad364e35_10.0.18362.449_none_6e1ee4e4ec7c6dc6/f/combase.dll
/cygdrive/c/Windows/WinSxS/wow64_microsoft-windows-com-base_31bf3856ad364e35_10.0.18362.449_none_6e1ee4e4ec7c6dc6/r/combase.dll

**Comment by mintty** [comment]
My previous comment was written before your's before (race condition); let's forget about anti-virus for now. Looking forward to some tracing results.

**Comment by retry** [comment]
OK *phew*.  I get worried when the finger pointing begins.  It's happened to me so many times *glares at freenode*.  I won't be able to attempt to compile mintty until later today.  I appreciate your attention on this.

**Comment by retry** [comment]
Cygwin has a mintty-debuginfo package you can install. Do you reckon that'd help as much as a self-compile?

**Comment by retry** [comment]
forget I asked. mintty.exe.dbg just segfaults.  I'll compile from a fresh github download.

**Comment by retry** [comment]
~/mintty-3.1.0/bin $ ./mintty.exe
begin debug
calling win_fix_taskbar_max
calling win_get_pixels
calling child_create
**1 minute hang occurs here**
calling ShowWindow
calling SetFocus

**Comment by mintty** [comment]
What's mintty.exe.dbg? I don't have experience with gdb but you could `gdb mintty`, set a breakpoint to functions to measure the time of arrive at them (e.g. `break child_create`) and start with `run -d` (the -d is important here). This may give you a glimpse of insight each time; tracing printfs are more targeted.

**Comment by mintty** [comment]
Excellent. So let's repeat this inside child_create.
Particularly, please trace before `forkpty` and after it, the latter including the pid value (to tell apart the output from parent and child process).

**Comment by retry** [comment]
Let's not worry about the cygwin debug packages.  So I went into child.c and started peppering that with printf's too:

~/mintty-3.1.0 $ bin/mintty.exe
begin debug
calling win_fix_taskbar_max
calling win_get_pixels
calling child_create
Child Create called
setting up signals
calling forkpty
** the long hang **
pid is 21345
I am the parent process
Calling fctnl xxx
if cfg.create_utmp
opening /dev/windows
calling ShowWindow
calling SetFocus

so your guess that it has to do with the forking looks correct!  Now what :D ?

**Comment by retry** [comment]
OH and then inside the child (the new mintty window)
pid is 0
Resetting signals
Terminal line settings
Calling execvp
~/mintty-3.1.0 $

But I'll go back and dig into the forkpty call more...

**Comment by BrianInglis** [comment]
You could try starting mintty running under strace e.g. shortcut c:\cygwin\bin\strace -o mintty.strace c:\cygwin\bin\mintty ... and email mintty.strace as a text attachment to an email with a suitable subject to cygwin@cygwin.com for diagnosis. 
Unfortunately strace only tells you where in the Cygwin interface things stopped working, and GDB does not work well with Cygwin Windows GUI programs, so rebuilding with added diagnostic output is the only good alternative. 
If mintty is now making network calls, Comodo Firewall is now likely hooking itself into or in place of multiple DLLs in the call chain, and it is known to interfere with Cygwin programs.

**Comment by mintty** [comment]
forkpty is a POSIX library call. There is a disabled self-made version of it in std.c, once intended for a very old cygwin version t hat did not have forkpty. You could try to activate it (change the #ifdef, rename functions and invocation...) to see whether it makes any difference.... probably not. Which would bring me to the point I'd have to give up, sorry. Cygwin experts may have further ideas.

**Comment by mintty** [comment]
Brian may still be right. Even if you disabled Comodo actions, it is still there and may be hooking in, who knows. You can only be sure if you completely uninstall it.

**Comment by retry** [comment]
Well I DID alter the code to use the built-in forkpty (renamed it forkptyx and set the cygwin version cutoff to < 9999 for that block of functions in std.c).  I still get the long hang, but mintty did load with this obsolete fork function.  WHat is weird though is the hang happens while calling the function!

in child.c:
~~~
printf("calling forkptyx\n");
  // Create the child process and pseudo terminal.
  pid = forkptyx(&pty_fd, 0, 0, winp);
~~~

then in std.c:
~~~
int forkptyx(int *amaster, char *name,
        const struct termios *termp, const struct winsize *winp)
{
printf("forkptyx called");
~~~

and I'm getting the hang between those two printf's with no other code between them that I can see.

**Comment by BrianInglis** [comment]
Another issue may be the default frequency of update checking is high at 900 seconds - 15 minutes: a lower default of a day (86400s) or many hours would seem to be more reasonable and be less questionable behaviour, especially for laptop users on insecure and/or billed connections (just like in dialup days). 
I didn't realize this has been around for a couple of years, as I maintain options settings and docs in /etc/minttyrc and ~/.minttyrc and only browse the Options dialogues so ~/.minttyrc content does not get messed up. 
As I update only via Cygwin, I have set this to 0 in my ~/.minttyrc.

**Comment by BrianInglis** [comment]
Both Cygwin console and pty handling and mintty console/pty handling have changed recently and some patches have and are being reworked or backed out because of problems.

**Comment by mintty** [comment]
The delay between calling and entering forkptyx is truely puzzling and I doubt it can be explained with console or pty stuff. The only, rather clueless, idea I get is that the second output is getting delayed, not its printf invocation. Adding fflush(stdout) is probably useless. Replace it with exit(0) for a try maybe.

**Comment by mintty** [comment]
Brian, the update checking is only ever invoked while the Options dialog is open, so the 15min is not a frequence, more a limit.

**Comment by retry** [comment]
Thanks for the tip on running strace.  I ran it in the background and immediately fired up a tail -f on the log.  That allowed me to catch exactly where the hang point is.  I cut the log off around the point bash is the main activity.  
[mintty.strace.log](https://github.com/mintty/mintty/files/4136651/mintty.strace.log)

Line 718 is the hanging point.

**Comment by retry** [comment]
I know you guys are saying it's a cygwin problem, and that seems reasonable at this point.  I don't expect you to examine my strace log unless you really want to.  I wonder though. Can either of you think of any other cygwin standard tool that forks? I would love to see if this forking problem is systemic or unusual to mintty.

**Comment by BrianInglis** [comment]
The incremental and cumulative microsecond timestamps indicate something went somewhere for just over 60 seconds before the worker threads could be terminated, and mintty could start looking for cygwin-console-helper to handle native Windows programs: 

--- Process 3152 (pid: 23072) thread 10024 exited with status 0x0
--- Process 3152 (pid: 23072) thread 7408 exited with status 0x0
--- Process 3152 (pid: 23072) thread 2992 exited with status 0x0
60013802 60107414 [main] mintty 23072 normalize_posix_path: src /bin/cygwin-console-helper.exe

One of the problems monitors have with Cygwin is that it uses highly parallel threading and locks shared memory areas to control the back and forth between threads using them, and those happen in different processes during forking, so monitors appear to be unable to cope with the highly parallel threading and locking performed, either delaying or faulting the processing. 
It appears something may have given up after a minute and allowed processing to proceed. 

Any shell does a lot of forking but nothing else works with ptys other than X window terminals xterm, rxvt. 

I am on the latest Windows 10 feature release at home, and have never seen nor heard of any such issues with mintty, even in corporate environments where everything was monitored and logged, and the possibilities of issues were high. The main corporate issues were slowdowns due to domain controller access to Active Directory.

**Comment by mintty** [comment]
> nothing else works with ptys other than X window terminals xterm, rxvt.

Likely also `screen` and `tmux`; can you try any of these please?

**Comment by retry** [comment]
OK tmux has the same problem.  Windows reports that it's waiting on Network I/O in the wait chain.  It did request access to 127.0.0.1 early on, which I granted, but it continued to hang for a minute just the same.  A second run, with the firewall rule already set, yielded the same exact hanging problem, just like mintty.

**Comment by mintty** [comment]
Thanks for testing and reporting.
I suggest you take the discussion to the cygwin mailing list then. You may refer to this issue which I'll close now.
